protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolSession.java

```diff
@@ -148,10 +148,39 @@ public String toString() {
 
     
     /**
-     * Return the {@link InetSocketAddress} of the remote peer
+     * Return the {@link InetSocketAddress} of the remote peer. If proxy support
+     * is enabled, then it returns the remote peer given by the proxy.
      */
     InetSocketAddress getRemoteAddress();
 
+    /**
+     * Return true if PROXY is required by the configuration
+     *
+     * @return supported
+     */
+    boolean isProxyRequired();
+
+    /**
+     * Return the {@link InetSocketAddress} of the proxy peer or null if proxy support is disabled.
+     */
+    InetSocketAddress getProxyDestinationAddress();
+
+    /**
+     * Sets the {@link InetSocketAddress} of the proxy peer.
+     */
+    void setProxyDestinationAddress(InetSocketAddress addr);
+
+    /**
+     * Return the {@link InetSocketAddress} of the remote peer if proxy is enabled or null if proxy
+     * support is disabled.
+     */
+    InetSocketAddress getProxySourceAddress();
+
+    /**
+     * Sets the {@link InetSocketAddress} remote peer provided by the proxy.
+     */
+    void setProxySourceAddress(InetSocketAddress addr);
```

Comment:
I do not like this API change much as...

 - 1. We do not know that a null value might be returned. NPEs are waiting to happen
 - 2. Nothing forbids me to set the proxy source and not the destination, or vice versa.
 
 Let me suggest to...
 
  - Create a `ProxyInformation` POJO
  
```
class ProxyInformation {
    private final InetSocketAddress source;
    private final InetSocketAddress destination;
    
    // constructor, equals, hash code, getters...
```

Then in protocol session we can... 

```
    void setProxyInformation(ProxyInformation info);
    Optional<ProxyInformation> getProxyInformation();
```

No more null, no more partially specified proxy info.

WDYT ?

Reply:
Had the same thought first, but I wanted it to keep simple. I will do so.